Optional fetch with credentials

mcclurec opened the pull request:
## Description

Follow up to [PR 291](https://github.com/bastienwirtz/homer/pull/291). Adding `credentials: 'include'` to all fetch calls had unexpected CORS side effects for users. This PR introduces a global mixin to expose a `fetchOptions` helper method on all components. "Service" items can now include a `fetchWithCredentials` prop on their config entry which will add the fetch option as needed.

Fixes # Issues [296](https://github.com/bastienwirtz/homer/issues/296) and [297](https://github.com/bastienwirtz/homer/issues/297)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I've read & comply with the [contributing guidelines](https://github.com/bastienwirtz/homer/blob/main/CONTRIBUTING.md)
- [x] I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers. 
- [x] I have made corresponding changes the documentation (README.md).
- [x] I've checked my modifications for any breaking changes, especially in the `config.yml` file

D4rkiii reviewed (approved):
LGTM.

One question. Maybe we should reconsider in the next major version how to define some header values?
Maybe one option on each service with the name "headers" and you can set your own values. It would be totally generic but for each service the same instead each service provides own options like apiKey...

Thanks for this update!

bastienwirtz commented:
Hi there ! 

@mcclurec I just pushed the [mixing](https://github.com/bastienwirtz/homer/blob/main/src/mixins/service.js) I had in mind, and refoactored [ping](https://github.com/bastienwirtz/homer/blob/main/src/components/services/Ping.vue) and [AdGuard](https://github.com/bastienwirtz/homer/blob/main/src/components/services/AdGuardHome.vue) service to use it (and all other services as soon as I find the time). 

What do you think? Let me know if you have any question (I will improve the documentation about that soon).

mcclurec commented:
@bastienwirtz I refactored my PR to utilize the `service` mixin's fetch. I removed the json [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) and the `endpoint` property in favor of a more transparent implementation of `fetch`.

Let me know if you have strong opinions about either of those choices. My thought is that it would be better to get this out in the community now and fix the CORS issues, and if we later want to refactor the mixin to handle more of the data fetching nitty gritty, we can do that in a separate PR.

bastienwirtz commented:
Hey @mcclurec,

The mixin is a shortcut, and even if it's 100% compatible with the native fetch signature, it's not meant to be a generic replacement. It provides the error checking and json decoding, which is useful for ALL services so far (with the boolean to handle specific case, and beyond that you the native fetch is still here if somethins), that's our use case here. It helps standardize all services implementation and ease the custom service development. 
So I rather keep the first implementation. I'll push the refactoring of the other services to use the internal fetch asap.

bastienwirtz commented:
All right, I finally pushed the last parts of the re-factorization to use the internal fetch method & the Generic template as a base. Credential include is now optional and service code is much smaller. We can close this issue. 
Thanks again @mcclurec for your work on this.
